onnxruntime/python/tools/transformers/convert_generation.py

```diff
@@ -1066,6 +1075,40 @@ def update_decoder_subgraph_past_present_share_buffer(subg):
     return subg
 
 
+def update_decoder_subgraph_use_decoder_masked_multihead_attention(subg) -> bool:
+    """Update the Attention nodes to DecoderMaskedMultiheadAttention.
+
+    Args:
+        subg (GraphProto): GraphProto of the decoder subgraph
+    """
+    decoder_masked_attention_supported_attr = [
+        "past_present_share_buffer",
+        "num_heads",
+        "scale",
+        "domain",
+    ]
+    new_nodes = []
+    for node in subg.node:
+        if node.op_type == "Attention":
```

Comment:
This conversion adds coupling of Attention op and DecoderMaskedMultiheadAttention op. That means we cannot change the interface of Attention alone. Maybe we shall document this (like Add a comment in Attention op spec so that we will not forget to update DecoderMaskedMultiheadAttention at the same time).

Another solution is to change fusion script to fuse DecoderMaskedMultiheadAttention instead of Attention. That's the reason we added an option in export:
https://github.com/microsoft/onnxruntime/blob/f88b97ede272e54d67e7f4da165359df4b534a4a/onnxruntime/python/tools/transformers/models/gpt2/convert_to_onnx.py#L164
Then in fusion, we can set an option in [FusionOptions](https://github.com/microsoft/onnxruntime/blob/f88b97ede272e54d67e7f4da165359df4b534a4a/onnxruntime/python/tools/transformers/fusion_options.py#L22), then the fusion could create DecoderMaskedMultiheadAttention or Attention accordingly based on the option:
https://github.com/microsoft/onnxruntime/blob/f88b97ede272e54d67e7f4da165359df4b534a4a/onnxruntime/python/tools/transformers/fusion_gpt_attention.py#L193
That could make it easier to test since there are tools for GPT-2 model, without testing beams search.

It's fine that this PR uses conversion approach. Just some suggestion that we might change it in the future.

Reply:
Good point. This blatantly couples Attention with  DecoderMaskedMultiheadAttention. Going through the solution you pasted below.

Reply:
Added a comment to the Attention op schema